### PR TITLE
feat: Implement TableProvider for Trait for `Db`

### LINKED ISSUE #701: feat: Implement DataFusion TableProvider for Chunks
__Use case:__

1. Allow for data streaming in to the bottom of the plan (avoid buffering all the data prior to conversion)
2. Allow the creation of generic datafusion plans for general purpose predicates (e.g. queries such as measurement names with arbitrary predicates)
3. Sets IOx up to use DataFusion optimizations (e.g. predicate pushdown and column projection)

This also avoids one extract copy of data while doing partition snap shotting :bonus: (it still buffers the entire parquet file in memory before sending to the object store)

Allowing DataFusion to see tables that are stored across chunks as "tables" is needed to use DataFusion across data stored in multiple chunks. The most recent time I  wanted this was as I was working on supporting general purpose predicate fallback in the `measurement_names` rpc call -- I need to generate a scan with arbitrary predicates and I found myself implementing something similar.

I actually think having a general purpose `Union` operator in DataFusion is probably the best eventually (and I might go add that shortly) but we will still need most of the same stuff in this PR for SQL anyways (where we need one table provider view across chunks and maybe even partitions).g

__Proposal:__
Short summary of the feature.

__Current behaviour:__
Describe what currently happens.

__Desired behaviour:__
Describe what you want.

__Alternatives considered:__
Describe other solutions or features you considered.

### PR DESCRIPTION
**Pull request by alamb**
Closes #701

## Rationale: 

TLDR is that this is needed to supporting streaming execution, selection and predicate pushdown into the Mutable Buffer, Read Buffer and Parquet chunks. 

### New Features
This PR adds the following features itself
* Streams reads out of the ReadBuffer and MutableBuffer during (SQL) queries
* Projection pushdown (only materialize columns needed by query rather than all of them) during (SQL) queries
* Streams data during the creation of parquet snapshots

Streaming in this case means "process one record batch at time" rather than "get them all at once". This is beneficial so that data isn't buffered overly long as well as limiting wasted work if the plan shuts down early (eg. `.. LIMIT 10`) if results aren't sent

While I specifically say "SQL" above, the mechanism will work for all DataFusion based plans (which will be everything except some metadata queries that are special cased)

### Not (yet) Implemented

This PR allows for, but does not yet implement, the following features (I am going to track them in their own tasks)

2. Implement Predicate pushdown in `TableProvider`
3. Implement chunk pruning across chunks of the read buffer, mutable buffer and parquet files
4. Implement partition pruning optimization to rule out data in a partition based on predicates and partition key
1. Implement statistics for `TableProvider` (let DataFusion planner know about cardinalities, etc).

## Changes in this PR

1. Replaces the "give me all your data" `table_to_arrow` call in the `PartitionCunk` with `read_filter` which takes a predicate and selection and does a proper read (with a filter!!!)
2. Implements the `TableProvider` trait and adapters that allow for streaming results.

After this change, you can no longer run SQL queries against the mutable buffer directly (you have to do so via the `Db` object), which is a step towards getting query logic out of the mutable buffer. 🎉 

## A note on testing
As this PR is mostly wiring up a bunch of the DataFusion APIs -- there is almost no new logic (and thus now new tests, as the features are covered via the end to end query tests in db.rs).

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: mutable_buffer/src/database.rs [hunk 1]
```diff
@@ -1012,30 +1009,6 @@ mod tests {
         Ok(())
     }
 
-    #[tokio::test]
-    async fn write_and_query() -> Result {
-        let db = MutableBufferDb::new("foo");
```
Comment (alamb) [inline mutable_buffer/src/database.rs:1017]: SQL can no longer be run directly against the mutable buffer -- by design, we are moving to something that works across all chunk types. Thus remove this test

File: query/src/frontend/sql.rs [hunk 2]
```diff
@@ -62,41 +90,46 @@ impl SQLQueryPlanner {
...
+            .map_err(|e| Box::new(e) as _)
+            .context(GettingDatabasePartition)?;
 
         // Register a table provider for each table so DataFusion
         // knows what the schema of that table is and how to obtain
         // its data when needed.
-        for table in &table_names {
-            let mut data = Vec::new();
+        for table_name in &table_names {
+            let mut builder = ProviderBuilder::new(table_name);
```
Comment (alamb) [inline query/src/frontend/sql.rs:103]: Here is the where the table provider is plumbed in

File: query/src/provider.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    /// returns Ok(combined_schema) if the schema of chunk is compatible with
+    /// `existing_schema`, Err() with why otherwise
+    fn check_schema(
+        &self,
+        existing_schema: SchemaRef,
+        chunk_schema: SchemaRef,
+    ) -> Result<SchemaRef> {
+        // For now, use strict equality. Eventually should union the schema
+        if existing_schema != chunk_schema {
```
Comment (alamb) [inline query/src/provider.rs:88]: I just noticed this here -- I will also file a ticket to implement proper schema merging here (we already have the logic to do so in `data_types` -- it just needs to be wired up.

File: query/src/provider.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        self.schema.clone()
+    }
+
+    fn scan(
+        &self,
+        projection: &Option<Vec<usize>>,
+        _batch_size: usize,
+        _filters: &[Expr],
+    ) -> std::result::Result<Arc<dyn ExecutionPlan>, DataFusionError> {
+        // TODO Here is where predicate pushdown will happen.  To make
```
Comment (alamb) [inline query/src/provider.rs:154]: Note here is the hook for proper predicate pushdown

File: query/src/provider/physical.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+        Ok(Arc::new(new_self))
+    }
+
+    async fn execute(
+        &self,
+        partition: usize,
+    ) -> arrow_deps::datafusion::error::Result<SendableRecordBatchStream> {
+        let fields = self.schema.fields();
+        let selection_cols = self.projection.as_ref().map(|projection| {
```
Comment (alamb) [inline query/src/provider/physical.rs:89]: The handling of selections across `await` boundaries is somewhat cumbersome -- I have some thoughts on how to make it better in the future

File: server/src/db/chunk.rs [hunk 6]
```diff
@@ -253,4 +210,82 @@ impl PartitionChunk for DBChunk {
...
+            DBChunk::ReadBuffer {
+                db,
+                partition_key,
+                chunk_id,
+            } => {
+                let chunk_id = *chunk_id;
+                let rb_predicate =
+                    to_read_buffer_predicate(&predicate).context(InternalPredicateConversion)?;
+                let db = db.read().expect("mutex poisoned");
+                let read_results = db
```
Comment (alamb) [inline server/src/db/chunk.rs:245]: here is the interface to the read buffer

File: server/src/snapshot.rs [hunk 7]
```diff
@@ -151,16 +164,22 @@ where
 
     async fn run(&self, notify: Option<oneshot::Sender<()>>) -> Result<()> {
         while let Some((pos, table_name)) = self.next_table() {
-            let mut batches = Vec::new();
-            self.partition
```
Comment (alamb) [inline server/src/snapshot.rs:155]: this used to buffer twice -- once for all the record batches, and once into a `Vec<u8>` -- now it just buffers into the `Vec<u8>` which still isn't ideal but it is progress

File: server/src/snapshot.rs [hunk 8]
```diff
@@ -354,29 +385,26 @@ cpu,host=B,region=east user=10.0,system=74.1 1
...
         "#;
 
-        let lines: Vec<_> = parse_lines(lp).map(|l| l.unwrap()).collect();
-        let write = lines_to_replicated_write(1, 1, &lines, &DatabaseRules::default());
-        let mut chunk = ChunkWB::new(11);
-
-        for e in write.write_buffer_batch().unwrap().entries().unwrap() {
-            chunk.write_entry(&e).unwrap();
-        }
+        let db = make_db();
```
Comment (alamb) [inline server/src/snapshot.rs:388]: changes these tests so they use a `Db` rather than `MutableBuffer` directly

File: server/src/db/chunk.rs [hunk 9]
```diff
@@ -253,4 +210,82 @@ impl PartitionChunk for DBChunk {
...
+
+                Ok(Box::pin(ReadFilterResultsStream::new(read_results, schema)))
+            }
+            DBChunk::ParquetFile => {
+                unimplemented!("parquet file not implemented for scan_data")
+            }
+        }
+    }
+
+    fn could_pass_predicate(&self, _predicate: &Predicate) -> Result<bool> {
```
Comment (e-dard) [inline server/src/db/chunk.rs:268]: Is this supposed to determine whether to bother issuing a read operation against the buffer? I'm wondering if this is different to actually issuing the read and getting no results back.

File: server/src/db/streams.rs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    }
+}
+
+impl RecordBatchStream for MutableBufferChunkStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+impl futures::Stream for MutableBufferChunkStream {
```
Comment (e-dard) [inline server/src/db/streams.rs:104]: So I'm thinking this is implementing an async stream from an iterator of results right?

**Review by e-dard — APPROVED** [review]
This is very cool 👍

File: server/src/db/chunk.rs (hunk 9 above)
Comment (alamb) [reply server/src/db/chunk.rs:268]: > Is this supposed to determine whether to bother issuing a read operation against the buffer? I'm wondering if this is different to actually issuing the read and getting no results back.

Comment (alamb) [reply server/src/db/chunk.rs:268]: > Is this supposed to determine whether to bother issuing a read operation against the buffer? 

Yes

> I'm wondering if this is different to actually issuing the read and getting no results back.

For the read buffer, it may not be different. I think for things like Parquet files it may be different indeed (aka don't even bother opening the files, etc)

File: server/src/db/streams.rs (hunk 10 above)
Comment (alamb) [reply server/src/db/streams.rs:104]: Yes, that is exactly what it is doing 😄

File: query/src/provider.rs (hunk 3 above)
Comment (alamb) [reply query/src/provider.rs:88]: https://github.com/influxdata/influxdb_iox/issues/732

**Comment by alamb** [comment]
Follow on work tracked in tickets:

* Ticket for table statistics: https://github.com/influxdata/influxdb_iox/issues/733
* Ticket for implementing predicate pushdown: https://github.com/influxdata/influxdb_iox/issues/734
* Ticket for DB chunk pruning: https://github.com/influxdata/influxdb_iox/issues/735
* Ticket for partition pruning: https://github.com/influxdata/influxdb_iox/issues/736

**Review by carols10cents — COMMENTED** [review]
Errrr I had started a review, but I just noticed this is now merged, feel free to ignore or implement my comments in future PRs I guess!

**Comment by alamb** [comment]
Thanks @carols10cents  -- I will implement these suggestions as a follow on PR.

File: query/src/frontend/sql.rs [hunk 11]
```diff
@@ -31,14 +29,44 @@ pub enum Error {
...
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display(
+        "Error adding chunk to table provider for table '{}' in chunk {}: {}",
+        table_name,
+        chunk_id,
+        source
+    ))]
+    AddingProviderToChunk {
```
[the remark this answers is not in the document]
Comment (alamb) [reply query/src/frontend/sql.rs:55]: good catch

File: query/src/provider.rs [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        chunk_schema
+    ))]
+    ChunkSchemaNotCompatible {
+        table_name: String,
+        existing_schema: SchemaRef,
+        chunk_schema: SchemaRef,
+    },
+
+    #[snafu(display(
+        "Internal error: no chunks found in builder for table {:?}",
```
[the remark this answers is not in the document]
Comment (alamb) [reply query/src/provider.rs:40]: yes, definitely. Good call

File: query/src/provider.rs [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        table_name,
+    ))]
+    NoChunks { table_name: String },
+
+    #[snafu(display("No rows found in table {}", table_name))]
+    InternalNoRowsInTable { table_name: String },
```
[the remark this answers is not in the document]
Comment (alamb) [reply query/src/provider.rs:46]: Yes. Excellent poiunt

**Comment by alamb** [comment]
Follow on PR: https://github.com/influxdata/influxdb_iox/pull/737
